### ISSUE TITLE
More sqrts than you can shake a stick at

### ISSUE THREAD

**Issue by kfsone** [shady831213/algorithms#1]
During your TSP you continually call sqrt while calculating distances. This is fairly wasteful, since you don't actually care about the root until the end of the function.

```
func dist2(lhs Point, rhs Point) float64 {
  dx, dy := lhs.x - rhs.x, lhs.y - rhs.y
  return (dx*dx) + (dy * dy)
}

func find_closest_2(points []Point) (float64, *Point, *Point, error) {
  if len(points) < 2 {
    return 0, nil, nil, errors.New("must be at least 2 points")
  }
  var first, second *Point
  var shortest_dist float64 = math.MaxFloat64
  for lhs := 0; lhs < len(points) - 1; lhs++ {
    for rhs := lhs + 1; rhs < len(points); rhs++ {
      d2 := dist2(lhs, rhs)
      if d2 < shortest_dist {
        first, second = &lhs, &rhs
        shortest_dist = d2
      }
    }
  }
  return math.sqrt(shortest_dist), first, second, nil
}
```

**Comment by shady831213** [comment]
@kfsone Thanks! I admit my solutions is far away from best， you can send a PR add your solution and test if you want. I appreciate it :) Thank you very much!

**Comment by kfsone** [comment]
I'll get the pull request in this weekendSent from my Verizon, Samsung Galaxy smartphone
-------- Original message --------From: Apollo Li <notifications@github.com> Date: 6/2/19  6:23 PM  (GMT-08:00) To: shady831213/algorithms <algorithms@noreply.github.com> Cc: Oliver Smith <oliver@kfs.org>, Mention <mention@noreply.github.com> Subject: Re: [shady831213/algorithms] More sqrts than you can shake a stick at (#1) Closed #1.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/shady831213/algorithms/issues/1?email_source=notifications\u0026email_token=AACO3QILLMROOLLG7ACBWODPYRW75A5CNFSM4HPY5FQ2YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGORYIX6FQ#event-2383511318",
"url": "https://github.com/shady831213/algorithms/issues/1?email_source=notifications\u0026email_token=AACO3QILLMROOLLG7ACBWODPYRW75A5CNFSM4HPY5FQ2YY3PNVWWK3TUL52HS4DFWZEXG43VMVCXMZLOORHG65DJMZUWGYLUNFXW5KTDN5WW2ZLOORPWSZGORYIX6FQ#event-2383511318",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]
